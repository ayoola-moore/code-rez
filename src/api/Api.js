//TODO that off mock api
const API = {
    // recipes: 'http://www.mocky.io/v2/5c85f7a1340000e50f89bd6c',
    // ingredient: 'https://www.mocky.io/v2/5cb016f83100006c00e131b0',
    recipes: 'https://api.myjson.com/bins/7ckd0',
    ingredient: 'https://api.myjson.com/bins/yqi6s',
}

export default API
