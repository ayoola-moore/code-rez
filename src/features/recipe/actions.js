export const MAKE_RECIPE: string = 'MAKE_RECIPE'
export const FETCH_RECIPES: string = 'FETCH_RECIPES'
export const FETCH_INGREDIENTS: string = 'FETCH_INGREDIENTS'
export const RECOMMENDATIONS: string = 'RECOMMENDATIONS'

//Flow type check for actions
export type FetchRecipes = {
    type: typeof FETCH_RECIPES,
    recipes: Array<Object>,
}

export type FetchIngredients = {
    type: typeof FETCH_INGREDIENTS,
    ingredients: Object,
}

export type Recommendations = {
    type: typeof RECOMMENDATIONS,
    recommended: Array<Object>,
}

// Redux actions
export const fetchApi = () => ({
    type: MAKE_RECIPE,
})

export const fetchRecipes = (recipes: Array<Object>): FetchRecipes => ({
    type: FETCH_RECIPES,
    recipes,
})

export const fetchIngredients = (ingredients: Object): FetchIngredients => ({
    type: FETCH_INGREDIENTS,
    ingredients,
})

export const recommendations = (
    recommended: Array<Object>
): Recommendations => ({
    type: RECOMMENDATIONS,
    recommended,
})
