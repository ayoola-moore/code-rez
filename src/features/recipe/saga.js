import { takeLatest, call, put, all } from 'redux-saga/effects'
import type { Saga } from 'redux-saga'
import axios from 'axios'
import api from '../../api/Api'
import { groupBy } from '../../utils/arrayToObj'
import { useBeforeFilter } from '../../utils/recipeFilters'
import * as actions from './actions'

export const callApi = (url: string): any => axios.get(url)

export function* fetchSaga(): Generator<any, any, any> {
    yield takeLatest(actions.MAKE_RECIPE, fetchRecipes)

    const recipes = yield fetchRecipes()
    const ingredients = yield fetchIngredients()

    yield all([
        put(actions.fetchRecipes(recipes)),
        put(actions.fetchIngredients(ingredients)),
    ])
    yield* updateRecipe(recipes, ingredients)
}

export function* fetchRecipes(): Saga<void> {
    try {
        const recipes = yield call(callApi, api.recipes)
        return recipes.data.recipes
    } catch (e) {
        console.log('error', e)
    }
}

export function* fetchIngredients(): Saga<void> {
    try {
        const ingredientArray = yield call(callApi, api.ingredient)
        const ingredients = groupBy(ingredientArray.data.ingredients, 'title')
        return ingredients
    } catch (e) {
        console.log('error', e)
    }
}

export function* updateRecipe(
    recipe: Array<Object>,
    ingredients: Object
): Saga<void> {
    try {
        const recommend = useBeforeFilter(recipe, ingredients)
        const sortedRecommendations = recommend.sort((a, b) => {
            return b.filterId - a.filterId
        })
        yield put(actions.recommendations(sortedRecommendations)) // to change
    } catch (e) {
        console.log('error', e)
    }
}
