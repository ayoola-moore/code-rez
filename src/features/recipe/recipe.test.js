// I could break the test into individual files, but ...

import * as actions from './actions'
import recipeReducer from './reducers'
import { callApi } from './saga'
import api from '../../api/Api'
import { call, put } from 'redux-saga/effects'

// Redux action test
describe('Redux action test cases', () => {
    it('should fetch api:: fetchApi', () => {
        expect(actions.fetchApi()).toMatchSnapshot()
    })
    it('should fetch recipes:: fetchRecipes', () => {
        expect(actions.fetchRecipes()).toMatchSnapshot()
    })
    it('should fetch ingredients:: fetchIngredient', () => {
        expect(actions.fetchIngredients()).toMatchSnapshot()
    })
    it('should recommend based on filter:: recommendations', () => {
        expect(actions.recommendations()).toMatchSnapshot()
    })
})

describe('Redux state test cases', () => {
    it('should find recipes', () => {
        const action = {
            type: actions.fetchRecipes,
            recipes: [
                {
                    title: 'Ham and Cheese Toastie',
                    ingredients: ['Ham', 'Cheese', 'Bread', 'Butter'],
                },
            ],
        }
        expect(recipeReducer(undefined, action)).toMatchSnapshot()
    })

    it('should find ingredients', () => {
        const action = {
            type: actions.fetchIngredients,
            ingredients: [
                {
                    title: 'Ham and Cheese Toastie',
                    ingredients: ['Ham', 'Cheese', 'Bread', 'Butter'],
                },
            ],
        }
        expect(recipeReducer(undefined, action)).toMatchSnapshot()
    })

    it('should find recommendationas', () => {
        const action = {
            type: actions.recommendations,
            recommended: [],
        }
        expect(recipeReducer(undefined, action)).toMatchSnapshot()
    })
})

const gen = actions.fetchApi()

// Redux Saga test
describe('Saga flow check', () => {
    it.skip('should hit the api', () => {
        expect(gen.next().value).toEqual(call(callApi, api.recipes))
    })
    // more indepth test required, but its Easter Sunday on my computer :)
})
