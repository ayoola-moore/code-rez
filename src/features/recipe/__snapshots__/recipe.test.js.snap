// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Redux action test cases should fetch api:: fetchApi 1`] = `
Object {
  "type": "MAKE_RECIPE",
}
`;

exports[`Redux action test cases should fetch ingredients:: fetchIngredient 1`] = `
Object {
  "ingredients": undefined,
  "type": "FETCH_INGREDIENTS",
}
`;

exports[`Redux action test cases should fetch recipes:: fetchRecipes 1`] = `
Object {
  "recipes": undefined,
  "type": "FETCH_RECIPES",
}
`;

exports[`Redux action test cases should recommend based on filter:: recommendations 1`] = `
Object {
  "recommended": undefined,
  "type": "RECOMMENDATIONS",
}
`;

exports[`Redux state test cases should find ingredients 1`] = `
Object {
  "allRecipes": Array [],
  "ingredients": Object {},
  "recommended": Array [],
}
`;

exports[`Redux state test cases should find recipes 1`] = `
Object {
  "allRecipes": Array [],
  "ingredients": Object {},
  "recommended": Array [],
}
`;

exports[`Redux state test cases should find recommendationas 1`] = `
Object {
  "allRecipes": Array [],
  "ingredients": Object {},
  "recommended": Array [],
}
`;
